[42기 이동민] Conflict 해결

gmgmgun opened the pull request:
Add: Map 컴포넌트 추가

Add: .env 파일 추가

Modify: Map 컴포넌트 이름 (Map->KakaoMap) 변경
Add: Kakao Map 라이브러리(react-kakao-map-sdk) 추가

Modify: KakaoMap 컴포넌트 페이지별로 세분화, 각 페이지 기능 구현 후 병합할 예정
Add: 기능 테스트용 파일(KakaoMapParent.js) 추가
Add: Hotels 테이블 Mock Data 추가

Modify: KakaoMapHost 컴포넌트 이름 변경
Modify: KakaoMapMain 컴포넌트 마커->금액으로 변경(커스텀오버레이 사용)

Modify: Main 컴포넌트 수정(KakaoMap 컴포넌트 테스트용)

Modify: KakaoMapParent 컴포넌트 페이지별 Mock Data  다르게 받도록 수정

Modify: KakaoMap 상세 페이지 기능 일부 구현(호텔 위치 정보 받아와서 지도 상에 마킹)
Add: hotel.json 추가

Modify: KakaoMap 컴포넌트 마커 사이즈 수정

Add: divisioninfo.json 추가
Modify: KakaoMap 컴포넌트 HOC 구조로 변경|

Add: KakaoMapMainMarker 컴포넌트 추가
Modify: components 디렉토리 추가 및 구조 변경
Delete: KakaoMapParent 컴포넌트 삭제
Modify: props 구조분해할당 수정

Add: bootstrap 라이브러리 추가
@gmgmgun

## :: 최근 작업 주제 (하나 이상의 주제를 선택해주세요.)
- [x] 기능 추가
- [ ] 리뷰 반영
- [ ] 리팩토링
- [ ] 버그 수정
- [ ] 컨벤션 수정

<br />

## :: 구현 목표 (해당 브랜치에서 구현하고자 하는 하나의 목표를 설정합니다.)
카카오 맵 API를 활용하여 메인 페이지 및 상세 페이지, 호스트 페이지에 출력될 KakaoMap 컴포넌트 구현

<br />

## :: 구현 사항 설명 (작업한 내용을 상세하게 기록합니다.)
1. 카카오 맵 API와 연동
  - 공식 문서를 참조해 라이브러리를 설치하였습니다.
  - 등록된 JS app key로 카카오 맵 API와 연동하였습니다.

2. 메인 페이지 지도 구현
  - 실제 reponse 받을 데이터를 기반으로 Mock Data(/data/hotels.json)를 작성하였습니다.
  - fetch 메서드로 받은 Mock Data의 요소들을 map 메서드를 통해 지도 위에 표시할 마커로 변환하였습니다.
  - 라이브러리 내 커스텀 오버레이 기능을 이용해 기본 마커를 없애고 styled component로 스타일링하였습니다.
  - Mock Data 내 숙박 요금에 대한 값을 number 타입으로 변환하고, LocaleToString 메서드를 통해 포매팅하여 마커에 나타나도록 구현하였습니다. 

3. 상세 페이지 지도 구현
  - 실제 reponse 받을 데이터를 기반으로 Mock Data(/data/hotel.json)를 작성하였습니다.
  - data로 받은 객체 리터럴 내 latitude, longitude를 지도의 중앙값과 마커의 포지션값에 설정해주었습니다.
  - 커스텀 오버레이를 이용해 기본 마커 대신 새로운 이미지를 마커로 사용하였습니다.

4. 호스트 페이지 지도 구현 
  - input 창에 구 입력 시 해당 구 중심 좌표로 지도 및 마커 이동하도록 구현하였습니다.
  - 마커가 드래그가 가능하도록 구현하였습니다.
  - 마커가 이동하면 실시간으로 좌표값이 변경되어 저장하도록 구현하였습니다.


5. 컴포넌트 병합 
  - 병합 대신 KakaoMap 컴포넌트가 HOC로써 props (프롭명 type)에 따라 컴포넌트를 중계하도록 구현하였습니다.

6. 리팩토링이 필요한 부분
  - 통신 및 Map 컴포넌트가 마운트되는 컴포넌트에서 동작 확인 필요합니다.
  
<br />

## :: 성장 포인트 (해당 기능을 구현하며 고민했던 사항이나 새로 알게된 부분, 어려웠던 점 등을 작성합니다.)
- 카카오 맵 API를 이용하면서 써드 파티 라이브러리 사용하는 법을 배웠습니다.
- Styled Component를 활용해 스타일링하는 법을 익혔습니다.
- 기획 초기에는 처음부터 props에 따라 출력되는 지도의 옵션을 다르게 구현하려고 하였으나, 컴포넌트 재사용 경험이 적어 분기하는 로직을 구상하는데 어려움이 많았습니다. 그래서 페이지 별로 컴포넌트를 나누어 따로 따로 구현해본 후 한 컴포넌트에 병합하는 과정을 거치기로 하였습니다.
- 외부 라이브러리를 커스텀하는 방법에 대해 배웠습니다.
- 고차 컴포넌트를 이용해 하나의 컴포넌트로 다른 컴포넌트들을 중계하는 법을 사용해보았습니다.
- useRef 및 useCallback을 사용하면서 기존에 학습했던 hook 이외의 훅에 대한 사용법 및 개념을 배웠습니다.

<br />

## :: 기타 질문 및 특이 사항

gmgmgun committed:
Add: .env 파일 추가

sstaar91 commented on src/components/KakaoMap/KakaoMapDetail.js:
```diff
@@ -0,0 +1,36 @@
+import { Map, CustomOverlayMap } from 'react-kakao-maps-sdk';
+import styled from 'styled-components';
+
+const KakaoMapDetail = props => {
+  const { data, size } = props;
```
```suggestion
const KakaoMapDetail = ( { data, size } ) => {
```
이렇게 구조분해 할당 하는 방법도 있습니다~

gmgmgun replied:
피드백 반영하였습니다!

sstaar91 commented on src/components/KakaoMap/KakaoMapHost.js:
```diff
@@ -0,0 +1,157 @@
+  // }, []);
+
+  return (
+    <div style={{ width: '100%', height: '300px', position: 'relative' }}>
```
인라인 스타일은 최대한 지양해 주시는 것이 좋습니다.
해당 부부은 불가피한 상황이 아니라고 판단됩니다!

gmgmgun replied:
피드백 반영하였습니다!

sstaar91 reviewed (commented):
수고하셨습니다 동민님
수정 사항 반영해주세요!

gmgmgun committed:
Modify: package.json 수정

Add: Map 컴포넌트 추가

Add: .env 파일 추가

Modify: Map 컴포넌트 이름 (Map->KakaoMap) 변경
Add: Kakao Map 라이브러리(react-kakao-map-sdk) 추가

Modify: KakaoMap 컴포넌트 페이지별로 세분화, 각 페이지 기능 구현 후 병합할 예정
Add: 기능 테스트용 파일(KakaoMapParent.js) 추가
Add: Hotels 테이블 Mock Data 추가

Modify: KakaoMapHost 컴포넌트 이름 변경
Modify: KakaoMapMain 컴포넌트 마커->금액으로 변경(커스텀오버레이 사용)

Modify: Main 컴포넌트 수정(KakaoMap 컴포넌트 테스트용)

Modify: KakaoMapParent 컴포넌트 페이지별 Mock Data  다르게 받도록 수정

Modify: KakaoMap 상세 페이지 기능 일부 구현(호텔 위치 정보 받아와서 지도 상에 마킹)
Add: hotel.json 추가

Modify: KakaoMap 컴포넌트 마커 사이즈 수정

Add: divisioninfo.json 추가
Modify: KakaoMap 컴포넌트 HOC 구조로 변경

Add: KakaoMapMainMarker 컴포넌트 추가
Modify: components 디렉토리 추가 및 구조 변경
Delete: KakaoMapParent 컴포넌트 삭제
Modify: props 구조분해할당 수정

Add: bootstrap 라이브러리 추가
Modify: Conflict 해결

gmgmgun committed:
Modify: package.json 수정

sstaar91 reviewed (approved):
conflict 해결하고 rebase 해결하고 merge 해주세요
